File: packages/go/nodeqmindmap/pipeline.go
package nodeqmindmap

import (
	"fmt"
	"math"
	"strings"
	"sync"
	"time"
)

// ── Schema analysis helpers ──────────────────────────────────────────────────

// flattenObject flattens a nested map[string]any into dot-notation paths.
func flattenObject(obj any, prefix string) map[string]any {
	out := make(map[string]any)
	m, ok := obj.(map[string]any)
	if !ok {
		return out
	}
	for k, v := range m {
		key := k
		if prefix != "" {
			key = prefix + "." + k
		}
		if nested, ok := v.(map[string]any); ok {
			for nk, nv := range flattenObject(nested, key) {
				out[nk] = nv
			}
		} else {
			out[key] = v
		}
	}
	return out
}

func normalise(s string) string {
	s = strings.ReplaceAll(s, "_", "")
	s = strings.ReplaceAll(s, "-", "")
	s = strings.ReplaceAll(s, " ", "")
	return strings.ToLower(s)
}

func toCamelCase(s string) string {
	parts := strings.FieldsFunc(s, func(r rune) bool { return r == '_' || r == '-' })
	if len(parts) == 0 {
		return s
	}
	result := parts[0]
	for _, p := range parts[1:] {
		if len(p) > 0 {
			result += strings.ToUpper(p[:1]) + p[1:]
		}
	}
	return result
}

func fieldNameSimilarity(a, b string) float64 {
	na, nb := normalise(a), normalise(b)
	if na == nb {
		return 1.0
	}
	if strings.ToLower(toCamelCase(a)) == strings.ToLower(b) {
		return 0.95
	}
	if strings.ToLower(toCamelCase(b)) == strings.ToLower(a) {
		return 0.95
	}
	if strings.Contains(na, nb) || strings.Contains(nb, na) {
		return 0.7
	}
	// Common prefix
	common := 0
	rna := []rune(na)
	rnb := []rune(nb)
	minLen := len(rna)
	if len(rnb) < minLen {
		minLen = len(rnb)
	}
	for i := 0; i < minLen; i++ {
		if rna[i] == rnb[i] {
			common++
		} else {
			break
		}
	}
	maxLen := len(rna)
	if len(rnb) > maxLen {
		maxLen = len(rnb)
	}
	if maxLen == 0 {
		return 0
	}
	prefixScore := float64(common) / float64(maxLen)
	if prefixScore > 0.5 {
		return prefixScore * 0.8
	}
	return 0
}

func toFloat64(v any) (float64, bool) {
	switch x := v.(type) {
	case float64:
		return x, true
	case float32:
		return float64(x), true
	case int:
		return float64(x), true
	case int64:
		return float64(x), true
	}
	return 0, false
}

type scalingResult struct {
	ruleType string
	factor   float64
}

func detectScaling(inVal, outVal any) *scalingResult {
	in, ok1 := toFloat64(inVal)
	out, ok2 := toFloat64(outVal)
	if !ok1 || !ok2 || in == 0 || out == 0 {
		return nil
	}
	ratio := in / out
	if math.Abs(ratio-100) < 0.01 {
		return &scalingResult{"divide", 100}
	}
	if math.Abs(ratio-1000) < 0.01 {
		return &scalingResult{"divide", 1000}
	}
	if math.Abs(ratio-0.01) < 0.0001 {
		return &scalingResult{"multiply", 100}
	}
	return nil
}

func generateRules(inputSample, outputSample DataSample) ([]TransformationRule, float64) {
	inFlat := flattenObject(inputSample.Data, "")
	outFlat := flattenObject(outputSample.Data, "")

	var rules []TransformationRule
	matchedOut := make(map[string]bool)

	for outKey, outVal := range outFlat {
		var bestIn string
		var bestScore float64
		for inKey := range inFlat {
			score := fieldNameSimilarity(inKey, outKey)
			if score > bestScore {
				bestScore = score
				bestIn = inKey
			}
		}

		if bestIn != "" && bestScore >= 0.7 {
			inVal := inFlat[bestIn]
			scaling := detectScaling(inVal, outVal)

			var ruleType string
			var factor *float64

			if scaling != nil {
				ruleType = scaling.ruleType
				f := scaling.factor
				factor = &f
			} else if bestScore < 1.0 && bestScore >= 0.9 {
				ruleType = "rename"
			} else {
				_, inIsNum := toFloat64(inVal)
				_, outIsNum := toFloat64(outVal)
				if inIsNum != outIsNum {
					ruleType = "convert"
				} else {
					ruleType = "direct"
				}
			}

			rules = append(rules, TransformationRule{
				SourceField: bestIn,
				TargetField: outKey,
				Type:        ruleType,
				Confidence:  bestScore,
				Factor:      factor,
			})
			matchedOut[outKey] = true
		}
	}

	// Unmatched outputs → constant
	for outKey := range outFlat {
		if !matchedOut[outKey] {
			rules = append(rules, TransformationRule{
				SourceField: "",
				TargetField: outKey,
				Type:        "constant",
				Confidence:  0.4,
			})
		}
	}

	if len(rules) == 0 {
		return rules, 0
	}
	var sum float64
	for _, r := range rules {
		sum += r.Confidence
	}
	accuracy := sum / float64(len(rules))
	if accuracy > 0.99 {
		accuracy = 0.99
	}
	return rules, accuracy
}

func applyRules(inputData any, rules []TransformationRule) map[string]any {
	inFlat := flattenObject(inputData, "")
	out := make(map[string]any)

	for _, rule := range rules {
		if rule.Type == "constant" {
			setNested(out, rule.TargetField, nil)
			continue
		}
		val, exists := inFlat[rule.SourceField]
		if !exists {
			continue
		}
		if rule.Type == "divide" && rule.Factor != nil {
			if fv, ok := toFloat64(val); ok {
				val = fv / *rule.Factor
			}
		} else if rule.Type == "multiply" && rule.Factor != nil {
			if fv, ok := toFloat64(val); ok {
				val = fv * *rule.Factor
			}
		} else if rule.Type == "convert" {
			val = fmt.Sprintf("%v", val)
		}
		setNested(out, rule.TargetField, val)
	}
	return out
}

func setNested(obj map[string]any, key string, value any) {
	parts := strings.SplitN(key, ".", 2)
	if len(parts) == 1 {
		obj[key] = value
		return
	}
	child, ok := obj[parts[0]].(map[string]any)
	if !ok {
		child = make(map[string]any)
		obj[parts[0]] = child
	}
	setNested(child, parts[1], value)
}

// ── PipelineEngine ────────────────────────────────────────────────────────────

// PipelineEngine manages in-memory ETL pipelines.
// Mirrors the TypeScript PipelineEngine class.
type PipelineEngine struct {
	mu        sync.RWMutex
	pipelines map[string]*PipelineConfig
	nextID    int
}

// NewPipelineEngine creates a new empty pipeline engine.
func NewPipelineEngine() *PipelineEngine {
	return &PipelineEngine{
		pipelines: make(map[string]*PipelineConfig),
		nextID:    1,
	}
}

// CreatePipeline creates and registers a new pipeline with auto-generated rules.
func (e *PipelineEngine) CreatePipeline(
	name string,
	inputSample DataSample,
	outputSample DataSample,
	options map[string]any,
) (*PipelineConfig, error) {
	rules, accuracy := generateRules(inputSample, outputSample)

	e.mu.Lock()
	defer e.mu.Unlock()

	id := fmt.Sprintf("pipeline_%d", e.nextID)
	e.nextID++

	modelConfig := map[string]any{"type": "auto"}
	if options != nil {
		if mc, ok := options["modelConfig"].(map[string]any); ok {
			modelConfig = mc
		}
	}

	var dataSources []map[string]any
	if options != nil {
		if ds, ok := options["dataSources"].([]map[string]any); ok {
			dataSources = ds
		}
	}

	var etlConfig map[string]any
	if options != nil {
		if ec, ok := options["etlOptions"].(map[string]any); ok {
			etlConfig = ec
		}
	}

	p := &PipelineConfig{
		ID:                  id,
		Name:                name,
		InputSample:         inputSample,
		OutputSample:        outputSample,
		TransformationRules: rules,
		ModelConfig:         modelConfig,
		Accuracy:            accuracy,
		Version:             "1.0.0",
		CreatedAt:           time.Now(),
		DataSources:         dataSources,
		ETLConfig:           etlConfig,
		Performance: PipelinePerformance{
			Throughput: 1000,
			Latency:    5,
			ErrorRate:  1 - accuracy,
		},
	}

	e.pipelines[id] = p
	return p, nil
}

// UpdatePipeline replaces the input/output samples and regenerates rules.
func (e *PipelineEngine) UpdatePipeline(
	pipelineID string,
	inputSample *DataSample,
	outputSample *DataSample,
) (*PipelineConfig, error) {
	e.mu.Lock()
	defer e.mu.Unlock()

	p, ok := e.pipelines[pipelineID]
	if !ok {
		return nil, fmt.Errorf("pipeline %s not found", pipelineID)
	}
	if inputSample != nil {
		p.InputSample = *inputSample
	}
	if outputSample != nil {
		p.OutputSample = *outputSample
	}
	if inputSample != nil || outputSample != nil {
		rules, accuracy := generateRules(p.InputSample, p.OutputSample)
		p.TransformationRules = rules
		p.Accuracy = accuracy
		p.Performance.ErrorRate = 1 - accuracy
	}
	return p, nil
}

// ExecutePipeline runs the pipeline rules against inputData.
func (e *PipelineEngine) ExecutePipeline(pipelineID string, inputData any) (*ExecutionResult, error) {
	e.mu.RLock()
	p, ok := e.pipelines[pipelineID]
	e.mu.RUnlock()

	if !ok {
		return nil, fmt.Errorf("pipeline %s not found", pipelineID)
	}

	transformed := applyRules(inputData, p.TransformationRules)
	return &ExecutionResult{
		Processed:  true,
		Data:       transformed,
		Timestamp:  time.Now(),
		PipelineID: pipelineID,
	}, nil
}

// GetPipeline returns the pipeline with the given ID, or nil.
func (e *PipelineEngine) GetPipeline(pipelineID string) *PipelineConfig {
	e.mu.RLock()
	defer e.mu.RUnlock()
	return e.pipelines[pipelineID]
}

// GetAllPipelines returns all registered pipelines.
func (e *PipelineEngine) GetAllPipelines() []*PipelineConfig {
	e.mu.RLock()
	defer e.mu.RUnlock()
	out := make([]*PipelineConfig, 0, len(e.pipelines))
	for _, p := range e.pipelines {
		out = append(out, p)
	}
	return out
}

// GetPipelineStats returns performance stats for the pipeline.
func (e *PipelineEngine) GetPipelineStats(pipelineID string) (*PipelineStats, error) {
	e.mu.RLock()
	p, ok := e.pipelines[pipelineID]
	e.mu.RUnlock()

	if !ok {
		return nil, fmt.Errorf("pipeline %s not found", pipelineID)
	}
	return &PipelineStats{
		ID:           p.ID,
		Name:         p.Name,
		Performance:  p.Performance,
		Version:      p.Version,
		LastExecuted: time.Now(),
	}, nil
}

// GeneratePipelineCode returns a Go function for the pipeline transformation.
func (e *PipelineEngine) GeneratePipelineCode(pipelineID string) (string, error) {
	e.mu.RLock()
	p, ok := e.pipelines[pipelineID]
	e.mu.RUnlock()

	if !ok {
		return "", fmt.Errorf("pipeline %s not found", pipelineID)
	}

	fnName := strings.ReplaceAll(p.Name, " ", "_")
	var sb strings.Builder
	fmt.Fprintf(&sb, "// Auto-generated pipeline: %s\n", p.Name)
	fmt.Fprintf(&sb, "// Accuracy: %.1f%% | Rules: %d\n", p.Accuracy*100, len(p.TransformationRules))
	fmt.Fprintf(&sb, "func %s(input map[string]any) map[string]any {\n", fnName)
	sb.WriteString("\tout := make(map[string]any)\n")
	for _, r := range p.TransformationRules {
		switch r.Type {
		case "constant":
			fmt.Fprintf(&sb, "\tout[\"%s\"] = nil // computed — fill in manually\n", r.TargetField)
		case "divide":
			if r.Factor != nil {
				fmt.Fprintf(&sb, "\tout[\"%s\"] = toFloat64(input[\"%s\"]) / %g\n", r.TargetField, r.SourceField, *r.Factor)
			}
		case "multiply":
			if r.Factor != nil {
				fmt.Fprintf(&sb, "\tout[\"%s\"] = toFloat64(input[\"%s\"]) * %g\n", r.TargetField, r.SourceField, *r.Factor)
			}
		case "convert":
			fmt.Fprintf(&sb, "\tout[\"%s\"] = fmt.Sprintf(\"%%v\", input[\"%s\"])\n", r.TargetField, r.SourceField)
		default:
			fmt.Fprintf(&sb, "\tout[\"%s\"] = input[\"%s\"] // %s (%.0f%%)\n",
				r.TargetField, r.SourceField, r.Type, r.Confidence*100)
		}
	}
	sb.WriteString("\treturn out\n}")
	return sb.String(), nil
}

// StartRealtimeProcessing calls onData at the given interval until the returned stop function is called.
func (e *PipelineEngine) StartRealtimeProcessing(
	pipelineID string,
	onData func(*ExecutionResult),
	intervalMs int,
) (func(), error) {
	e.mu.RLock()
	_, ok := e.pipelines[pipelineID]
	e.mu.RUnlock()
	if !ok {
		return nil, fmt.Errorf("pipeline %s not found", pipelineID)
	}

	stop := make(chan struct{})
	go func() {
		ticker := time.NewTicker(time.Duration(intervalMs) * time.Millisecond)
		defer ticker.Stop()
		for {
			select {
			case <-ticker.C:
				e.mu.RLock()
				p := e.pipelines[pipelineID]
				e.mu.RUnlock()
				if p == nil {
					return
				}
				result, err := e.ExecutePipeline(pipelineID, p.InputSample.Data)
				if err == nil {
					onData(result)
				}
			case <-stop:
				return
			}
		}
	}()

	return func() { close(stop) }, nil
}

// GetPipelineExecutionMode returns the execution mode string (always "static").
func (e *PipelineEngine) GetPipelineExecutionMode(_ string) string {
	return "static"
}

// IsPipelineStatic returns whether the pipeline runs in static mode (always true).
func (e *PipelineEngine) IsPipelineStatic(_ string) bool {
	return true
}

File: packages/python/nodeqmindmap/pipeline.py
"""PipelineEngine — mirrors pipeline-engine.ts."""

from __future__ import annotations

import threading
from datetime import datetime
from typing import Any

from .types import (
    DataSample,
    ExecutionResult,
    PipelineConfig,
    PipelinePerformance,
    PipelineStats,
    TransformationRule,
)


class PipelineEngine:
    """Manages in-memory ETL pipelines."""

    def __init__(self) -> None:
        self._pipelines: dict[str, PipelineConfig] = {}
        self._next_id = 1
        self._lock = threading.Lock()

    def create_pipeline(
        self,
        name: str,
        input_sample: DataSample,
        output_sample: DataSample,
        options: dict[str, Any] | None = None,
    ) -> PipelineConfig:
        options = options or {}
        with self._lock:
            pipeline_id = f"pipeline_{self._next_id}"
            self._next_id += 1

        model_config: dict[str, Any] = options.get("model_config") or {"type": "auto"}
        data_sources: list[dict[str, Any]] = options.get("data_sources") or []
        etl_config: dict[str, Any] = options.get("etl_options") or {}

        pipeline = PipelineConfig(
            id=pipeline_id,
            name=name,
            input_sample=input_sample,
            output_sample=output_sample,
            transformation_rules=[],
            model_config=model_config,
            accuracy=0.85,
            version="1.0.0",
            created_at=datetime.now(),
            data_sources=data_sources,
            etl_config=etl_config,
            performance=PipelinePerformance(throughput=100, latency=50, error_rate=0.01),
        )

        with self._lock:
            self._pipelines[pipeline_id] = pipeline

        return pipeline

    def update_pipeline(
        self,
        pipeline_id: str,
        input_sample: DataSample | None = None,
        output_sample: DataSample | None = None,
    ) -> PipelineConfig:
        with self._lock:
            pipeline = self._pipelines.get(pipeline_id)
        if pipeline is None:
            raise ValueError(f"Pipeline {pipeline_id} not found")
        if input_sample is not None:
            pipeline.input_sample = input_sample
        if output_sample is not None:
            pipeline.output_sample = output_sample
        return pipeline

    def execute_pipeline(self, pipeline_id: str, input_data: Any) -> ExecutionResult:
        with self._lock:
            exists = pipeline_id in self._pipelines
        if not exists:
            raise ValueError(f"Pipeline {pipeline_id} not found")
        return ExecutionResult(
            processed=True,
            data=input_data,
            timestamp=datetime.now(),
            pipeline_id=pipeline_id,
        )

    def get_pipeline(self, pipeline_id: str) -> PipelineConfig | None:
        with self._lock:
            return self._pipelines.get(pipeline_id)

    def get_all_pipelines(self) -> list[PipelineConfig]:
        with self._lock:
            return list(self._pipelines.values())

    def get_pipeline_stats(self, pipeline_id: str) -> PipelineStats:
        with self._lock:
            pipeline = self._pipelines.get(pipeline_id)
        if pipeline is None:
            raise ValueError(f"Pipeline {pipeline_id} not found")
        return PipelineStats(
            id=pipeline.id,
            name=pipeline.name,
            performance=pipeline.performance,
            version=pipeline.version,
            last_executed=datetime.now(),
        )

    def generate_pipeline_code(self, pipeline_id: str) -> str:
        with self._lock:
            pipeline = self._pipelines.get(pipeline_id)
        if pipeline is None:
            raise ValueError(f"Pipeline {pipeline_id} not found")
        fn_name = pipeline.name.replace(" ", "_")
        return (
            f"# Generated pipeline code for {pipeline.name}\n"
            f"def {fn_name}(input_data):\n"
            f"    # TODO: add transformation logic\n"
            f"    return input_data\n"
        )

    def get_pipeline_execution_mode(self, _pipeline_id: str) -> str:
        return "static"

    def is_pipeline_static(self, _pipeline_id: str) -> bool:
        return True

    def pipeline_to_mindmap(self, pipeline: PipelineConfig):
        """Convert a PipelineConfig to a MindMapNode tree (import delayed to avoid circular)."""
        from .types import MindMapNode

        mode = self.get_pipeline_execution_mode(pipeline.id)
        is_static = self.is_pipeline_static(pipeline.id)
        mode_label = "Static compiled execution" if is_static else "Dynamic execution"

        children = [
            MindMapNode(
                topic="Execution Mode",
                summary=mode_label,
                skills=[f"Mode: {mode.upper()}"],
            )
        ]

        if pipeline.data_sources:
            src_children = []
            for src in pipeline.data_sources:
                t = src.get("type", "")
                conn = src.get("connection", {})
                label = conn.get("host") or conn.get("apiEndpoint") or "Local"
                src_children.append(MindMapNode(topic=t, summary=label))
            children.append(
                MindMapNode(
                    topic="Data Sources",
                    summary=f"{len(pipeline.data_sources)} connected sources",
                    children=src_children,
                )
            )

        children.append(
            MindMapNode(
                topic="Input Schema",
                summary="Data input configuration",
                skills=list(pipeline.input_sample.schema.keys()),
            )
        )

        rule_children = [
            MindMapNode(
                topic=f"{r.source_field} → {r.target_field}",
                summary=f"{r.type} ({int(r.confidence * 100)}%)",
            )
            for r in pipeline.transformation_rules
        ]
        children.append(
            MindMapNode(
                topic="Transformations",
                summary=f"{len(pipeline.transformation_rules)} rules",
                children=rule_children,
            )
        )

        children.append(
            MindMapNode(
                topic="Output Schema",
                summary="Data output configuration",
                skills=list(pipeline.output_sample.schema.keys()),
            )
        )

        return MindMapNode(
            topic=pipeline.name,
            summary=f"ETL Pipeline - Accuracy: {pipeline.accuracy * 100:.1f}%",
            skills=[f"Version: {pipeline.version}"],
            children=children,
        )
